shenanigan/trainers/core.py

```diff
@@ -40,26 +51,36 @@ def __call__(self, train_loader, val_loader, num_epochs):
             num_epochs: int
                 Number of epochs to train the model for.
         """
-        self.checkpointer.restore(use_pretrained=self.use_pretrained)
+        # NOTE the order is important, initialise model with save every as this is likely the last model saved
+        self.save_best_checkpointer.restore(use_pretrained=self.use_pretrained)
+        self.save_every_checkpointer.restore(use_pretrained=self.use_pretrained)
+
+        self.minimum_loss = self.save_best_checkpointer.get_loss()
 
         for _ in range(num_epochs):
             # Train
-            self.checkpointer.increment_epoch()
-            train_metrics = self.train_epoch(train_loader, self.checkpointer.get_epoch_num())
-            train_metrics['epoch'] = int(self.checkpointer.get_epoch_num())
+            self.save_every_checkpointer.increment_epoch()
+            self.save_best_checkpointer.increment_epoch()
+            train_metrics = self.train_epoch(train_loader, self.save_every_checkpointer.get_epoch_num())
+            train_metrics['epoch'] = int(self.save_every_checkpointer.get_epoch_num())
             print(f'Metrics: {train_metrics}')
             self.train_logger(train_metrics)
             # Validation
-            val_metrics = self.val_epoch(val_loader, self.checkpointer.get_epoch_num())
-            val_metrics['epoch'] = int(self.checkpointer.get_epoch_num())
+            val_metrics = self.val_epoch(val_loader, self.save_every_checkpointer.get_epoch_num())
+            val_metrics['epoch'] = int(self.save_every_checkpointer.get_epoch_num())
             print(f'Metrics: {val_metrics}')
             self.val_logger(val_metrics)
+            # update loss
+            self.save_best_checkpointer.update_loss(train_metrics[self.tracking_metric])
+            self.save_every_checkpointer.update_loss(train_metrics[self.tracking_metric])
             # Save
-            if ((self.checkpointer.get_epoch_num() + 1) % self.save_every == 0) or \
-               ((self.checkpointer.get_epoch_num() + 1) > self.save_best_after and self.is_best(val_metrics['generator_loss'])):
-                save_path = self.checkpointer.save()
-                print("Saved checkpoint for step {}: {}".format(int(self.checkpointer.get_epoch_num()), save_path))
-            self.run_callbacks(self.checkpointer.get_epoch_num())
+            if ((self.save_every_checkpointer.get_epoch_num() + 1) % self.save_every) == 0:
```

Comment:
Perhaps we want the epoch number to start at 1 so that we don't have to remember to `+1` everywhere?